src/Sentry/Internal/CachingBackgroundWorker.cs

```diff
@@ -0,0 +1,146 @@
+                catch (IOException ex)
+                {
+                    // Don't delete the cache file in case of transient exceptions,
+                    // i.e. loss of connection, failure to connect, etc.
+                    // Instead break to not violate the order and wait until
+                    // the next batch to attempt sending again.
+                    Options.DiagnosticLogger?.Log(
+                        SentryLevel.Error,
+                        "Transient failure when sending cached envelope: {0}",
+                        ex,
+                        envelope.TryGetEventId()
+                    );
+
+                    break;
+                }
+                catch (Exception ex)
+                {
+                    // Delete the cache file for all other exceptions that could
+                    // indicate a successfully completed, but error response.
+                    envelopeFile.Delete();
+
+                    Options.DiagnosticLogger?.Log(
+                        SentryLevel.Error,
+                        "Persistent failure when sending cached envelope: {0}",
+                        ex,
+                        envelope.TryGetEventId()
+                    );
+                }
+            }
+        }
+
+        private async ValueTask<FileInfo> StoreEnvelopeAsync(
+            Envelope envelope,
+            CancellationToken cancellationToken = default)
+        {
+            // If over capacity - remove oldest envelope file
+            // TODO: probably a good idea to put a lock here to make sure this limit is maintained
+            while (GetEnvelopeFiles().Count() >= 30)
+            {
+                if (TryGetNextEnvelopeFile() is { } oldestEnvelopeFile)
+                {
+                    oldestEnvelopeFile.Delete();
+                }
+            }
+
+            // Envelope file name can be either:
+            // 1604679692_b2495755f67e4bb8a75504e5ce91d6c1_17754019.envelope
+            // 1604679692__17754019.envelope
+            // (depending on whether event ID is present or not)
+            var envelopeFile = new FileInfo(
+                Path.Combine(
+                    _cacheDirectory.FullName,
+                    $"{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}_" +
+                    $"{envelope.TryGetEventId()}_" +
+                    $"{envelope.GetHashCode()}" +
+                    $".{EnvelopeFileExt}")
+            );
+
+            _cacheDirectory.Create();
+            using var stream = envelopeFile.Create();
+            await envelope.SerializeAsync(stream, cancellationToken).ConfigureAwait(false);
+
+            return envelopeFile;
+        }
+
+        protected override async ValueTask ProcessEnvelopeAsync(
+            Envelope envelope,
+            CancellationToken cancellationToken = default)
+        {
+            // Add this envelope to cache
+            await StoreEnvelopeAsync(envelope, cancellationToken).ConfigureAwait(false);
```

Comment:
If we throw here, the worker thread above is supposed to log and continue, right?
I wonder if we'll drop the item and deal with it "gracefully".
We need to make sure we don't end up retrying something broken in a tight loop here on a worker thread or killing that thread and stopping all event submission

Reply:
Yeah, this method is called from the worker which has like 5 layers of `try`/`catch` :)
How do you think the workflow should look like?

Reply:
> Yeah, this method is called from the worker which has like 5 layers of try/catch :)


Does this mean the worker won't crash? I believe that's the case.

Reply:
Yeah it won't crash.